File: LeetCode75_LevelOne/Number_of_Islands/Number_of_Islands.py
# 200. Number of Islands Python

class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.island(grid, i, j)
                    count += 1
        return count
    def island(self, grid, i, j):
        if i < 0 or j < 0 or i > len(grid) - 1 or j > len(grid[0]) - 1 or grid[i][j] != '1':
            return
        grid[i][j] = '&'
        self.island(grid, i-1, j)
        self.island(grid, i+1, j)
        self.island(grid, i, j-1)
        self.island(grid, i, j+1)
        

File: LeetCode75_LevelOne/Number_of_Islands/Number_of_Islands.cc
// 200. Number of Islands C++

#include <vector>

class Solution {
public:
    int rows, cols;
    int numIslands(std::vector<std::vector<char>>& grid) {
        int count = 0;
        if (grid.empty()) return count;
        rows = grid.size();
        cols = grid[0].size();
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1') {
                    island(grid, i, j);
                    count++;
                }
            }
        }
        return count;
    }
    void island(std::vector<std::vector<char>>& grid, int i, int j) {
        if (i < 0 || j < 0 || i >= rows || j >= cols || grid[i][j] != '1') return;
        grid[i][j] = '&';
        island(grid, i-1, j);
        island(grid, i+1, j);
        island(grid, i, j-1);
        island(grid, i, j+1);
    }
};
